api/segments/models.py

```diff
@@ -109,9 +115,14 @@ def id_exists_in_rules_data(rules_data: typing.List[dict]) -> bool:
     def does_identity_match(
         self, identity: "Identity", traits: typing.List["Trait"] = None
     ) -> bool:
-        rules = self.rules.all()
-        return rules.count() > 0 and all(
-            rule.does_identity_match(identity, traits) for rule in rules
+        segment_model = map_segment_to_engine(self)
+        identity_model = map_identity_to_engine(identity)
+        trait_models = map_traits_to_trait_models(traits) if traits else None
```

Comment:
My concern here is that it's a bit hidden as to what additional database interactions are happening. I think, to plugin the engine, we need to do it much earlier in the flow and use e.g. `engine.get_identity_feature_states`.

Reply:
> additional database interactions

These are well covered by the currently failing `assertNumQueries` assertion in `test_get_segments`. IMO if we manage to keep the same amount (or less) by prefetching the data, we're good.

> do it much earlier in the flow and use e.g. engine.get_identity_feature_states.

I'll think of it but my primary focus here is to replace the segment matching logic first.

Reply:
>IMO if we manage to keep the same amount (or less) by prefetching the data, we're good.

Yep, fair enough.